src/node_usb.cc

```diff
@@ -12,32 +10,68 @@ Napi::Value RefHotplugEvents(const Napi::CallbackInfo& info);
 Napi::Value UnrefHotplugEvents(const Napi::CallbackInfo& info);
 void initConstants(Napi::Object target);
 
-libusb_context* usb_context;
-struct HotPlug {
-	libusb_device* device;
-	libusb_hotplug_event event;
-};
+void handleHotplug(HotPlug* info){
+	Napi::ObjectReference* hotplugThis = info->hotplugThis;
+	Napi::Env env = hotplugThis->Env();
+	Napi::HandleScope scope(env);
+
+	libusb_device* dev = info->device;
+	libusb_hotplug_event event = info->event;
+
+	DEBUG_LOG("HandleHotplug %p %i", dev, event);
+
+	Napi::Value v8dev = Device::get(env, dev);
+	libusb_unref_device(dev);
 
-std::thread usb_thread;
+	Napi::String eventName;
+	if (LIBUSB_HOTPLUG_EVENT_DEVICE_ARRIVED == event) {
+		DEBUG_LOG("Device arrived");
+		eventName = Napi::String::New(env, "attach");
+
+	} else if (LIBUSB_HOTPLUG_EVENT_DEVICE_LEFT == event) {
+		DEBUG_LOG("Device left");
+		eventName = Napi::String::New(env, "detach");
 
-void USBThreadFn(){
+	} else {
+		DEBUG_LOG("Unhandled hotplug event %d\n", event);
+		return;
+	}
+
+	hotplugThis->Get("emit").As<Napi::Function>().MakeCallback(hotplugThis->Value(), { eventName, v8dev });
+	delete info;
+}
+
+int LIBUSB_CALL hotplug_callback(libusb_context* ctx, libusb_device* device,
+                     libusb_hotplug_event event, void* user_data) {
+	libusb_ref_device(device);
+	ModuleData* instanceData = (ModuleData*)user_data;
+	instanceData->hotplugQueue.post(new HotPlug {device, event, &instanceData->hotplugThis});
+	return 0;
+}
+
+void USBThreadFn(Napi::Env env) {
+	libusb_context* usb_context = env.GetInstanceData<ModuleData>()->usb_context;
 	while(1) libusb_handle_events(usb_context);
 }
 
 Napi::Object Init(Napi::Env env, Napi::Object exports) {
 	Napi::HandleScope scope(env);
+	env.SetInstanceData(new ModuleData());
+	ModuleData* instanceData = env.GetInstanceData<ModuleData>();
 
 	initConstants(exports);
 
 	// Initialize libusb. On error, halt initialization.
+	libusb_context* usb_context;
```

Comment:
this should be initialized to `nullptr`, so that (if libusb_init fails) we don't store garbage into `usb_context` in line 67 below (this is so we don't call `libusb_exit` on garbage, see comment below)

Reply:
Done